### DISCUSSION TITLE
Sharing ccache and -march=native

### DISCUSSION THREAD

**Discussion by ghost** [ccache/ccache#823] [Support]
Suppose I share a cache between two computers of the same architecture but different capabilities (say, x86_64, one has avx, one does not have avx).

The compiler in both machines is exactly the same binary.

I compile something with `-march=native` on the computer with avx, so that avx instructions are used.

Later I compile the exact same file, with the exact same compiler, with the exact same flags (including `-march=native`) on the computer without avx.

Then ccache sees everything is exactly the same (source, compiler, flags) and so it gives a hit and returns the exact same object file.

But the object file will NOT work in the computer without avx !!!

What are my alternatives? I guess the ideal is that when I use `-march=native`, the cache entry depends on having the exact same cpu flags (???).

Thanks a lot for your help.

**Comment by jrosdahl** [comment]
This is in a bug in ccache: it should include some architecture-specific information in the hash if `-march=native` is used. Not sure what that information should be, though. I filed a bug report for it: #824

Here are two different workarounds I can think of:

* Create (or find) an architecture-specific file and configure it as [`extra_files_to_hash`](https://ccache.dev/manual/4.2.html#config_extra_files_to_hash).
* Configure a [`compiler_check`](https://ccache.dev/manual/4.2.html#config_compiler_check) command that includes architecture-specific information in addition to the compiler version.

**Reply by ghost** [reply]
Thanks a lot for your answer. As a workaround I'm using
```
compiler_check = %compiler% -v ; hostname
```
which seems to work, but the effect is equivalent to not sharing the cache. I wonder:

a. Is there a way I could make the `compiler_check` command depend on the parameters passed to the compiler, so that I only append `hostname` when invoked with `-march=native` ?

b. The default for `compiler_check` is `mtime`. My first idea was using `compiler_check = mtime ; hostname`, which I think it's incorrect. But I really only wanted to append the output of `hostname` to the default. Is this possible?

Maybe a simple way to fix this would be to add a configuration option `compiler_check_native` which applies exactly when `-march=native` is among the parameters. In https://github.com/ccache/ccache/issues/824#issuecomment-803708213 I suggested
```
$ gcc -march=native -### -E - < /dev/null 2>&1 | sed -ne 's/.*cc1 .*-march=\([^ "]*\)[ "].*/\1/p'
nehalem
```
as a possible reasonable default for this for gcc, I don't know about clang.

Note that hostname is not reasonable --- too fine grained --- although in my case is equivalent.

**Reply by jrosdahl** [reply]
> a. Is there a way I could make the `compiler_check` command depend on the parameters passed to the compiler, so that I only append `hostname` when invoked with `-march=native` ?

Not currently. It would be quite easy to add a token that expands to the compiler arguments (e.g. `%arguments%`), though.

> b. The default for `compiler_check` is `mtime`. My first idea was using `compiler_check = mtime ; hostname`, which I think it's incorrect. But I really only wanted to append the output of `hostname` to the default. Is this possible?

No, it's either a command or one of the built-in methods.

> Maybe a simple way to fix this would be to add a configuration option `compiler_check_native` which applies exactly when `-march=native` is among the parameters.

That sounds a bit too special to me. If we add an `%arguments%` token you can write a small shell script that expresses the logic you want. Sounds reasonable?

**Reply by ghost** [reply]
> > a. Is there a way I could make the `compiler_check` command depend on the parameters passed to the compiler, so that I only append `hostname` when invoked with `-march=native` ?
> 
> Not currently. It would be quite easy to add a token that expands to the compiler arguments (e.g. `%arguments%`), though.
> 
> > b. The default for `compiler_check` is `mtime`. My first idea was using `compiler_check = mtime ; hostname`, which I think it's incorrect. But I really only wanted to append the output of `hostname` to the default. Is this possible?
> 
> No, it's either a command or one of the built-in methods.
> 
> > Maybe a simple way to fix this would be to add a configuration option `compiler_check_native` which applies exactly when `-march=native` is among the parameters.
> 
> That sounds a bit too special to me. If we add an `%arguments%` token you can write a small shell script that expresses the logic you want. Sounds reasonable?

Good enough, this would allow one to solve the issue.

It would be nice to have the issue solved by default out of the box (as is I have to know about the issue and configure around it). I don't know how to do it in a way that is correct for all compilers.

I'd prefer to do within the configuration file, not requiring an external shell script, and also that I can fall back to the default `mtime` when compiling without `-march`.

**Reply by jrosdahl** [reply]
> It would be nice to have the issue solved by default out of the box (as is I have to know about the issue and configure around it). I don't know how to do it in a way that is correct for all compilers.

Agree. That's what #824 is about.

> I'd prefer to do within the configuration file, not requiring an external shell script, and also that I can fall back to the default `mtime` when compiling without `-march`.

Right, relying on an external script is not optimal, but then again this is just a workaround for a real solution (the `compiler_check` setting is about identifying the compiler, not the architecture).

**Reply by ghost** [reply]
> > It would be nice to have the issue solved by default out of the box (as is I have to know about the issue and configure around it). I don't know how to do it in a way that is correct for all compilers.
> 
> Agree. That's what #824 is about.
> 
> > I'd prefer to do within the configuration file, not requiring an external shell script, and also that I can fall back to the default `mtime` when compiling without `-march`.
> 
> Right, relying on an external script is not optimal, but then again this is just a workaround for a real solution (the `compiler_check` setting is about identifying the compiler, not the architecture).

Sure! I'm sorry I sounded negative. Having `%arguments%` would be wonderful and it will enable a better workaround than what is possible now.

**Comment by ghost** [comment]
For the record: another command that might be suitable is
```
$ gcc -march=native -Q --help=target
The following options are target specific:
  -m128bit-long-double                  [disabled]
  -m16                                  [disabled]
  -m32                                  [disabled]
  -m3dnow                               [disabled]
  -m3dnowa                              [disabled]
  -m64                                  [enabled]
  -m80387                               [enabled]
  -m8bit-idiv                           [disabled]
  -m96bit-long-double                   [enabled]
  -mabi=                                sysv
  -mabm                                 [disabled]
  -maccumulate-outgoing-args            [disabled]
  -maddress-mode=                       short
  -madx                                 [disabled]
  -maes                                 [disabled]
  -malign-data=                         compat
  -malign-double                        [disabled]
  -malign-functions=                    0
  -malign-jumps=                        0
  -malign-loops=                        0
  -malign-stringops                     [enabled]
  -mandroid                             [disabled]
  -march=                               nehalem
  -masm=                                att
  -mavx                                 [disabled]
  -mavx2                                [disabled]
....
```

A difference with the full output of `gcc -march=native -### -E - < /dev/null 2>&1 | grep cc1` is that it does not include information about cache sizes.
